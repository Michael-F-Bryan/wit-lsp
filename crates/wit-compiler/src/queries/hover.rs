use either::Either;
use tree_sitter::Range;

use crate::{
    ast::{AstNode, HasSource},
    access::{
        EnumIndex, FlagsIndex, FuncItemIndex, GetAstNode, GetByIndex, Index, InterfaceIndex,
        RecordIndex, ResourceIndex, TypeAliasIndex, VariantIndex, WorldIndex,
    },
    queries::{FilePath, ItemDefinitionMetadata, Items, SourceFile},
    Db, Text, Tree,
};

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum HoverTarget {
    Interface(InterfaceIndex),
    World(WorldIndex),
    Enum(Either<InterfaceIndex, WorldIndex>, EnumIndex),
    Flags(Either<InterfaceIndex, WorldIndex>, FlagsIndex),
    Resource(Either<InterfaceIndex, WorldIndex>, ResourceIndex),
    Variant(Either<InterfaceIndex, WorldIndex>, VariantIndex),
    FuncItem(Either<InterfaceIndex, WorldIndex>, FuncItemIndex),
    TypeAlias(Either<InterfaceIndex, WorldIndex>, TypeAliasIndex),
    Record(Either<InterfaceIndex, WorldIndex>, RecordIndex),
}

#[salsa::tracked]
pub fn hover_info(db: &dyn Db, file: SourceFile, target: HoverTarget) -> HoverInfo {
    let items = crate::queries::file_items(db, file);
    let ast = crate::queries::parse(db, file);
    let tree = ast.tree(db);

    let state = State {
        db,
        file,
        items,
        tree,
    };

    match target {
        HoverTarget::Interface(index) => hover_interface(index, state),
        HoverTarget::World(index) => hover_world(index, state),
        HoverTarget::Enum(scope, index) => state.hover_type(scope, index),
        HoverTarget::Flags(scope, index) => state.hover_type(scope, index),
        HoverTarget::Resource(scope, index) => state.hover_type(scope, index),
        HoverTarget::Variant(scope, index) => state.hover_type(scope, index),
        HoverTarget::FuncItem(scope, index) => state.hover_type(scope, index),
        HoverTarget::TypeAlias(scope, index) => state.hover_type(scope, index),
        HoverTarget::Record(scope, index) => state.hover_type(scope, index),
    }
}

struct State<'db> {
    db: &'db dyn Db,
    file: SourceFile,
    items: Items,
    tree: &'db Tree,
}

impl<'db> State<'db> {
    fn source(&self, node: impl HasSource) -> &str {
        let src = self.file.contents(self.db);
        node.utf8_text(src)
    }

    fn hover_type<I>(&self, scope: Either<InterfaceIndex, WorldIndex>, index: I) -> HoverInfo
    where
        I: Index,
        ItemDefinitionMetadata: GetByIndex<I>,
        <ItemDefinitionMetadata as GetByIndex<I>>::Metadata: GetAstNode,
    {
        let ptr = match scope {
            Either::Left(interface) => {
                let world = self.items.get_by_index(self.db, interface);
                world.get_by_index(self.db, index)
            }
            Either::Right(world) => {
                let world = self.items.get_by_index(self.db, world);
                world.get_by_index(self.db, index)
            }
        };

        let node = ptr.ast_node(self.tree);

        HoverInfo {
            filename: self.file.path(self.db).clone(),
            definition: node.range(),
            content: self.source(node).into(),
        }
    }
}

fn hover_interface(index: InterfaceIndex, s: State<'_>) -> HoverInfo {
    let meta = s.items.get_interface(s.db, index);
    let node = meta.location(s.db).ast_node(s.tree);

    HoverInfo {
        filename: s.file.path(s.db).clone(),
        definition: node.range(),
        content: s.source(node).into(),
    }
}

fn hover_world(index: WorldIndex, s: State<'_>) -> HoverInfo {
    let meta = s.items.get_world(s.db, index);
    let node = meta.location(s.db).ast_node(s.tree);

    HoverInfo {
        filename: s.file.path(s.db).clone(),
        definition: node.range(),
        content: s.source(node).into(),
    }
}

/// Information shown when hovering over an item.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct HoverInfo {
    /// The name of the file this thing came from.
    pub filename: FilePath,
    /// Where the item was defined. Useful when you want to jump to definition.
    pub definition: Range,
    /// The contents of the tooltip, formatted as Markdown text.
    pub content: Text,
}
