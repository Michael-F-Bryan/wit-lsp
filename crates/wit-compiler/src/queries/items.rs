#![allow(clippy::too_many_arguments)]

use either::Either;
use im::{ordmap::Entry, OrdMap, Vector};
use tree_sitter::{Node, Point, Range};

use crate::{
    access::{
        ConstructorPtr, EnumIndex, EnumPtr, FlagsIndex, FlagsPtr, FuncItemIndex, FunctionPtr,
        GetAstNode, Index, InterfaceIndex, InterfacePtr, MethodPtr, Pointer, RawIndex, RecordIndex,
        RecordPtr, ResourceIndex, ResourceMethodIndex, ResourcePtr, StaticMethodPtr,
        StaticResourceMethodIndex, TypeAliasIndex, TypeAliasPtr, VariantIndex, VariantPtr,
        WorldIndex, WorldPtr,
    },
    ast::{self, AstNode, HasIdent},
    diagnostics::{Diagnostic, Diagnostics, Location},
    queries::SourceFile,
    Db, Text,
};

/// Parse a file and walk its AST to find the names of all the items it contains.
#[salsa::tracked]
#[tracing::instrument(level = "debug", skip_all, fields(file = %file.path(db)))]
pub fn file_items(db: &dyn Db, file: SourceFile) -> Items {
    let ast = crate::queries::parse(db, file);
    let src = ast.src(db);
    let root = ast.source_file(db);

    let mut names = NameTable::new(db, file);
    let mut worlds = Vector::new();
    let mut worlds_by_name = OrdMap::new();
    let mut interfaces = Vector::new();
    let mut interfaces_by_name = OrdMap::new();

    for top_level_item in root.iter_top_level_items() {
        let node = top_level_item.syntax();

        if let Some(world) = top_level_item.world_item() {
            if let Some((name, world)) = walk_world(db, world, src, file) {
                if names.insert(name.clone(), node) {
                    let ix = WorldIndex::from_raw(RawIndex::new(worlds.len()));
                    worlds.push_back(world);
                    worlds_by_name.insert(name, ix);
                }
            }
        } else if let Some(interface) = top_level_item.interface_item() {
            if let Some((name, interface)) = walk_interface(db, interface, src, file) {
                if names.insert(name.clone(), node) {
                    let ix = InterfaceIndex::from_raw(RawIndex::new(worlds.len()));
                    interfaces.push_back(interface);
                    interfaces_by_name.insert(name, ix);
                }
            }
        }
    }

    Items::new(db, worlds_by_name, worlds, interfaces_by_name, interfaces)
}

fn walk_world(
    db: &dyn Db,
    world: ast::WorldItem<'_>,
    src: &str,
    file: SourceFile,
) -> Option<(Text, WorldMetadata)> {
    let name = Text::from(world.identifier(src)?);
    let location = WorldPtr::for_node(world);

    let mut names = NameTable::new(db, file);

    let mut builder = TypesBuilder::new(src);

    for item in world.iter_items() {
        if let Some(typedef) = item.typedef_item() {
            builder.process(typedef, &mut names);
        }
    }

    let meta = WorldMetadata::new(db, name.clone(), location, builder.finish());

    Some((name, meta))
}

fn walk_interface(
    db: &dyn Db,
    interface: ast::InterfaceItem<'_>,
    src: &str,
    file: SourceFile,
) -> Option<(Text, InterfaceMetadata)> {
    let name = Text::from(interface.identifier(src)?);
    let location = InterfacePtr::for_node(interface);

    let mut names = NameTable::new(db, file);

    let mut builder = TypesBuilder::new(src);

    for item in interface.iter_items() {
        if let Some(f) = item.func_item() {
            builder.process_function(f, &mut names);
        } else if let Some(typedef) = item.typedef_item() {
            builder.process(typedef, &mut names);
        }
    }

    let meta = InterfaceMetadata::new(db, name.clone(), location, builder.finish());

    Some((name, meta))
}

#[derive(Debug, Default)]
struct TypesBuilder<'a> {
    src: &'a str,
    enums_by_name: OrdMap<Text, EnumIndex>,
    enums: Vector<EnumPtr>,
    flags_by_name: OrdMap<Text, FlagsIndex>,
    flags: Vector<FlagsPtr>,
    functions_by_name: OrdMap<Text, FuncItemIndex>,
    functions: Vector<FunctionPtr>,
    records_by_name: OrdMap<Text, RecordIndex>,
    records: Vector<RecordPtr>,
    resources_by_name: OrdMap<Text, ResourceIndex>,
    resources: Vector<ResourceMetadata>,
    typedefs_by_name: OrdMap<Text, TypeAliasIndex>,
    typedefs: Vector<TypeAliasPtr>,
    variants_by_name: OrdMap<Text, VariantIndex>,
    variants: Vector<VariantPtr>,
}

impl<'a> TypesBuilder<'a> {
    fn new(src: &'a str) -> Self {
        TypesBuilder {
            src,
            ..Default::default()
        }
    }

    fn process(&mut self, typedef: ast::TypedefItem<'a>, names: &mut NameTable<'_, 'a>) {
        if let Some(e) = typedef.enum_item() {
            insert(e, self.src, names, &mut self.enums, &mut self.enums_by_name);
        } else if let Some(v) = typedef.variant_item() {
            insert(
                v,
                self.src,
                names,
                &mut self.variants,
                &mut self.variants_by_name,
            );
        } else if let Some(r) = typedef.record_item() {
            insert(
                r,
                self.src,
                names,
                &mut self.records,
                &mut self.records_by_name,
            );
        } else if let Some(r) = typedef.resource_item() {
            let Some(name) = r.identifier(self.src) else {
                return;
            };

            let meta = process_resource(names.db, names.file, r, name);
            let name = meta.name.clone();

            if names.insert(name.clone(), r.syntax()) {
                let index = Index::from_raw(RawIndex::new(self.resources.len()));
                self.resources.push_back(meta);
                self.resources_by_name.insert(name, index);
            }
        } else if let Some(f) = typedef.flags_item() {
            insert(f, self.src, names, &mut self.flags, &mut self.flags_by_name);
        } else if let Some(t) = typedef.type_item() {
            insert(
                t,
                self.src,
                names,
                &mut self.typedefs,
                &mut self.typedefs_by_name,
            );
        }
    }

    fn finish(self) -> ItemDefinitionMetadata {
        let TypesBuilder {
            enums_by_name,
            enums,
            functions,
            functions_by_name,
            flags_by_name,
            flags,
            records_by_name,
            records,
            resources_by_name,
            resources,
            typedefs_by_name,
            typedefs,
            variants_by_name,
            variants,
            ..
        } = self;
        ItemDefinitionMetadata {
            enums_by_name,
            enums,
            flags_by_name,
            flags,
            functions_by_name,
            functions,
            records_by_name,
            records,
            resources_by_name,
            resources,
            typedefs_by_name,
            typedefs,
            variants_by_name,
            variants,
        }
    }

    fn process_function(&mut self, f: ast::FuncItem<'a>, names: &mut NameTable<'_, 'a>) {
        insert(
            f,
            self.src,
            names,
            &mut self.functions,
            &mut self.functions_by_name,
        );
    }
}

fn process_resource(
    db: &dyn Db,
    file: SourceFile,
    node: ast::ResourceItem<'_>,
    name: &str,
) -> ResourceMetadata {
    let location = ResourcePtr::for_node(node);
    let name = Text::from(name);
    let src = file.contents(db);
    let mut constructor: Option<ConstructorPtr> = None;

    let mut names = NameTable::new(db, file);

    let mut methods = Vector::new();
    let mut static_methods = Vector::new();
    let mut methods_by_name = OrdMap::new();
    let mut static_methods_by_name = OrdMap::new();

    for m in node.iter_methods() {
        if let Some(c) = m.resource_constructor() {
            if let Some(previous) = constructor {
                let location = Location::new(file.path(db), c.range());
                let original_location = Location::new(file.path(db), previous.range());
                let diag = Diagnostic::multiple_constructors(location, original_location);
                Diagnostics::push(db, diag);
                continue;
            }

            constructor = Some(ConstructorPtr::for_node(c));
        } else if let Some(f) = m.func_item() {
            if let Some(name) = f.identifier(src) {
                let name = Text::from(name);
                if names.insert(name.clone(), f.syntax()) {
                    let index = ResourceMethodIndex::from_raw(RawIndex::new(methods.len()));
                    methods_by_name.insert(name, index);
                    methods.push_back(MethodPtr::for_node(f));
                }
            }
        } else if let Some(s) = m.static_method() {
            if let Some(name) = s.identifier(src) {
                let name = Text::from(name);
                if names.insert(name.clone(), m.syntax()) {
                    let index = StaticResourceMethodIndex::from_raw(RawIndex::new(methods.len()));
                    static_methods_by_name.insert(name, index);
                    static_methods.push_back(StaticMethodPtr::for_node(s));
                }
            }
        }
    }

    ResourceMetadata {
        name,
        location,
        constructor,
        methods,
        methods_by_name,
        static_methods,
        static_methods_by_name,
    }
}

fn insert<'tree, Ast, Index, Ptr>(
    node: Ast,
    src: &str,
    names: &mut NameTable<'_, 'tree>,
    items: &mut Vector<Ptr>,
    by_name: &mut OrdMap<Text, Index>,
) where
    Ast: AstNode<'tree> + HasIdent + Copy,
    Index: crate::access::Index,
    Ptr: crate::access::Pointer<Node<'tree> = Ast> + Copy,
{
    let Some(name) = node.identifier(src) else {
        return;
    };

    let ptr = Ptr::for_node(node);

    let name = Text::from(name);

    if names.insert(name.clone(), node.syntax()) {
        let index = Index::from_raw(RawIndex::new(items.len()));
        items.push_back(ptr);
        by_name.insert(name, index);
    }
}

pub(crate) struct NameTable<'db, 'tree> {
    db: &'db dyn Db,
    file: SourceFile,
    names: OrdMap<Text, Node<'tree>>,
}

impl<'db, 'tree> NameTable<'db, 'tree> {
    pub fn new(db: &'db dyn Db, file: SourceFile) -> Self {
        NameTable {
            db,
            file,
            names: OrdMap::new(),
        }
    }

    pub fn insert(&mut self, name: Text, node: Node<'tree>) -> bool {
        match self.names.entry(name.clone()) {
            Entry::Vacant(entry) => {
                entry.insert(node);
                true
            }
            Entry::Occupied(entry) => {
                let diag = Diagnostic::duplicate_name(
                    entry.key().clone(),
                    Location::new(self.file.path(self.db), node.range()),
                    Location::new(self.file.path(self.db), entry.get().range()),
                );
                Diagnostics::push(self.db, diag);
                false
            }
        }
    }
}

/// Metadata about all items in a file.
#[salsa::tracked]
pub struct Items {
    pub worlds_by_name: OrdMap<Text, WorldIndex>,
    pub worlds: Vector<WorldMetadata>,
    pub interfaces_by_name: OrdMap<Text, InterfaceIndex>,
    pub interfaces: Vector<InterfaceMetadata>,
}

impl Items {
    pub fn iter_worlds(
        &self,
        db: &dyn Db,
    ) -> impl Iterator<Item = (Text, WorldIndex, WorldMetadata)> {
        let worlds = self.worlds(db);

        self.worlds_by_name(db)
            .into_iter()
            .map(move |(name, ix)| (name, ix, worlds[ix.raw().as_usize()]))
    }

    pub fn iter_interfaces(
        &self,
        db: &dyn Db,
    ) -> impl Iterator<Item = (Text, InterfaceIndex, InterfaceMetadata)> {
        let interfaces = self.interfaces(db);

        self.interfaces_by_name(db)
            .into_iter()
            .map(move |(name, ix)| (name, ix, interfaces[ix.raw().as_usize()]))
    }
}

#[salsa::tracked]
impl Items {
    pub fn get_world(&self, db: &dyn Db, index: WorldIndex) -> WorldMetadata {
        let worlds = self.worlds(db);
        worlds[index.raw().as_usize()]
    }

    pub fn get_interface(&self, db: &dyn Db, index: InterfaceIndex) -> InterfaceMetadata {
        let interfaces = self.interfaces(db);
        interfaces[index.raw().as_usize()]
    }

    pub fn enclosing_item(
        &self,
        db: &dyn Db,
        point: Point,
    ) -> Option<Either<WorldIndex, InterfaceIndex>> {
        for (i, meta) in self.worlds(db).into_iter().enumerate() {
            if range_contains(meta.location(db).range(), point) {
                return Some(Either::Left(WorldIndex::from_raw(RawIndex::new(i))));
            }
        }

        for (i, meta) in self.interfaces(db).into_iter().enumerate() {
            if range_contains(meta.location(db).range(), point) {
                return Some(Either::Right(InterfaceIndex::from_raw(RawIndex::new(i))));
            }
        }

        None
    }
}

fn range_contains(range: Range, point: Point) -> bool {
    range.start_point <= point && point <= range.end_point
}

#[salsa::tracked]
pub struct WorldMetadata {
    pub name: Text,
    pub location: WorldPtr,
    pub items: ItemDefinitionMetadata,
}

#[salsa::tracked]
pub struct InterfaceMetadata {
    pub name: Text,
    pub location: InterfacePtr,
    pub items: ItemDefinitionMetadata,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct ItemDefinitionMetadata {
    pub enums_by_name: OrdMap<Text, EnumIndex>,
    pub enums: Vector<EnumPtr>,
    pub flags_by_name: OrdMap<Text, FlagsIndex>,
    pub flags: Vector<FlagsPtr>,
    pub functions_by_name: OrdMap<Text, FuncItemIndex>,
    pub functions: Vector<FunctionPtr>,
    pub records_by_name: OrdMap<Text, RecordIndex>,
    pub records: Vector<RecordPtr>,
    pub resources_by_name: OrdMap<Text, ResourceIndex>,
    pub resources: Vector<ResourceMetadata>,
    pub typedefs_by_name: OrdMap<Text, TypeAliasIndex>,
    pub typedefs: Vector<TypeAliasPtr>,
    pub variants_by_name: OrdMap<Text, VariantIndex>,
    pub variants: Vector<VariantPtr>,
}

impl ItemDefinitionMetadata {
    pub fn iter_enums(&self) -> impl Iterator<Item = EnumIndex> + '_ {
        iter(&self.enums)
    }

    pub fn iter_flags(&self) -> impl Iterator<Item = FlagsIndex> + '_ {
        iter(&self.flags)
    }

    pub fn iter_functions(&self) -> impl Iterator<Item = FuncItemIndex> + '_ {
        iter(&self.functions)
    }

    pub fn iter_records(&self) -> impl Iterator<Item = RecordIndex> + '_ {
        iter(&self.records)
    }

    pub fn iter_resources(&self) -> impl Iterator<Item = ResourceIndex> + '_ {
        iter(&self.resources)
    }

    pub fn iter_typedefs(&self) -> impl Iterator<Item = TypeAliasIndex> + '_ {
        iter(&self.typedefs)
    }

    pub fn iter_variants(&self) -> impl Iterator<Item = VariantIndex> + '_ {
        iter(&self.variants)
    }

    pub fn names(&self) -> impl Iterator<Item = &Text> {
        self.enums_by_name
            .keys()
            .chain(self.flags_by_name.keys())
            .chain(self.functions_by_name.keys())
            .chain(self.records_by_name.keys())
            .chain(self.resources_by_name.keys())
            .chain(self.typedefs_by_name.keys())
            .chain(self.variants_by_name.keys())
    }
}

fn iter<Ix, Meta>(items: &Vector<Meta>) -> impl Iterator<Item = Ix> + '_
where
    Ix: Copy + Index,
    Meta: Clone,
{
    (0..items.len()).map(|ix| Ix::from_raw(RawIndex::new(ix)))
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct ResourceMetadata {
    pub name: Text,
    pub location: ResourcePtr,
    pub constructor: Option<ConstructorPtr>,
    pub methods_by_name: OrdMap<Text, ResourceMethodIndex>,
    pub methods: Vector<MethodPtr>,
    pub static_methods_by_name: OrdMap<Text, StaticResourceMethodIndex>,
    pub static_methods: Vector<StaticMethodPtr>,
}

impl ResourceMetadata {
    pub fn iter_methods(&self) -> impl Iterator<Item = (&Text, ResourceMethodIndex, MethodPtr)> {
        self.methods_by_name
            .iter()
            .map(|(name, ix)| (name, *ix, self.methods[ix.raw().as_usize()]))
    }

    pub fn iter_static_methods(
        &self,
    ) -> impl Iterator<Item = (&Text, StaticResourceMethodIndex, StaticMethodPtr)> {
        self.static_methods_by_name
            .iter()
            .map(|(name, ix)| (name, *ix, self.static_methods[ix.raw().as_usize()]))
    }
}

impl GetAstNode for ResourceMetadata {
    type Node<'tree> = ast::ResourceItem<'tree>;

    fn ast_node(self, tree: &crate::Tree) -> Self::Node<'_> {
        self.location.ast_node(tree)
    }
}
