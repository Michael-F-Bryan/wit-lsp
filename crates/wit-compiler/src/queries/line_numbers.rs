use std::ops::Range;

use codespan_reporting::files::Error as CodespanError;
use im::Vector;

use crate::{queries::SourceFile, Db, Text};

#[salsa::tracked]
pub fn calculate_line_numbers(db: &dyn Db, file: SourceFile) -> LineNumbers {
    let src = file.contents(db);
    LineNumbers::for_text(src.clone())
}

/// A lookup table for efficiently looking up line numbers in a file.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct LineNumbers {
    line_starts: Vector<usize>,
    src: Text,
}

impl LineNumbers {
    pub fn for_text(src: Text) -> Self {
        let line_starts = codespan_reporting::files::line_starts(&src).collect();
        LineNumbers { line_starts, src }
    }

    /// The index of the line at the given byte offset.
    ///
    /// If the byte index is past the end of the file, an error is returned.
    pub fn line_index(&self, byte_offset: usize) -> Result<usize, CodespanError> {
        if byte_offset > self.src.len() {
            return Err(CodespanError::IndexTooLarge {
                given: byte_offset,
                max: self.src.len(),
            });
        }

        match self.line_starts.binary_search(&byte_offset) {
            Ok(line) => Ok(line),
            Err(next_line) => Ok(next_line - 1),
        }
    }

    /// The byte range of a line in the source of the file.
    pub fn line_range(&self, line_index: usize) -> Result<Range<usize>, CodespanError> {
        let line_start = self.line_start(line_index)?;
        let next_line_start = self.line_start(line_index + 1)?;

        Ok(line_start..next_line_start)
    }

    /// Return the starting byte index of the line with the specified line index.
    pub fn line_start(&self, line_index: usize) -> Result<usize, CodespanError> {
        use std::cmp::Ordering;

        match line_index.cmp(&self.line_starts.len()) {
            Ordering::Less => Ok(self
                .line_starts
                .get(line_index)
                .cloned()
                .expect("failed despite previous check")),
            Ordering::Equal => Ok(self.src.len()),
            Ordering::Greater => Err(CodespanError::LineTooLarge {
                given: line_index,
                max: self.line_starts.len() - 1,
            }),
        }
    }

    pub fn point(&self, byte_index: usize) -> Result<tree_sitter::Point, CodespanError> {
        let row = self.line_index(byte_index)?;
        let range = self.line_range(row)?;
        let column = byte_index - range.start;

        Ok(tree_sitter::Point { row, column })
    }
}
