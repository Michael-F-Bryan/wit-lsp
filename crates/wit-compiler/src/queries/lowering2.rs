use crate::{
    hir,
    queries::{Package, Workspace},
    Db,
};

#[salsa::tracked]
#[tracing::instrument(level = "debug", skip_all, fields(dir = %pkg.dir(db).raw_path(db)))]
pub fn lower_package(db: &dyn Db, _ws: Workspace, pkg: Package) -> hir::Package {
    let _items = crate::queries::package_items(db, pkg);

    hir::Package {
        ..Default::default()
    }
}
