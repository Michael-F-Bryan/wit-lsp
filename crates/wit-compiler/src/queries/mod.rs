mod parsing;
mod selection;

pub use self::{
    parsing::{parse, Ast, Tree, Workspace},
    selection::selection_ranges,
};
