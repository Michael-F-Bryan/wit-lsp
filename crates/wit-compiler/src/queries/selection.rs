use im::Vector;
use tree_sitter::{Node, Point, Range};

use crate::{queries::Ast, Db};

/// Calculate successively wider ranges of tokens.
#[salsa::tracked]
pub fn selection_ranges(db: &dyn Db, ast: Ast, cursor_location: Point) -> Option<Vector<Range>> {
    selection_ranges_impl(ast.root_node(db), cursor_location)
}

fn selection_ranges_impl(root: Node<'_>, cursor_location: Point) -> Option<Vector<Range>> {
    if cursor_location > root.end_position() {
        // Out of bounds
        return None;
    }

    let mut cursor = root.walk();

    while cursor.goto_first_child_for_point(cursor_location).is_some() {
        // keep iterating
    }

    let mut ranges = Vector::new();
    ranges.push_back(cursor.node().range());

    while cursor.goto_parent() {
        let range = cursor.node().range();

        if ranges.last() == Some(&range) {
            // We want to ignore ranges that won't expand the selection.
            continue;
        }

        ranges.push_back(range);
    }

    Some(ranges)
}

#[cfg(test)]
mod tests {
    use crate::queries::Tree;

    use super::*;

    #[test]
    fn calculate_selection_ranges() {
        let src = include_str!("../../../../integration-tests/compile-pass/all-resources.wit");
        let tree = Tree::parse(src);
        let point = Point {
            row: 10,
            column: 18,
        };

        let ranges = selection_ranges_impl(tree.root_node(), point).unwrap();

        insta::assert_debug_snapshot!(ranges);
    }
}
