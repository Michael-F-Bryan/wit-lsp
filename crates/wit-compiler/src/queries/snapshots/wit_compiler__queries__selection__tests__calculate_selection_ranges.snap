---
source: crates/wit-compiler/src/queries/selection.rs
expression: ranges
---
[
    Range {
        start_byte: 123,
        end_byte: 124,
        start_point: Point {
            row: 10,
            column: 17,
        },
        end_point: Point {
            row: 10,
            column: 18,
        },
    },
    Range {
        start_byte: 122,
        end_byte: 128,
        start_point: Point {
            row: 10,
            column: 16,
        },
        end_point: Point {
            row: 10,
            column: 22,
        },
    },
    Range {
        start_byte: 121,
        end_byte: 129,
        start_point: Point {
            row: 10,
            column: 15,
        },
        end_point: Point {
            row: 10,
            column: 23,
        },
    },
    Range {
        start_byte: 110,
        end_byte: 130,
        start_point: Point {
            row: 10,
            column: 4,
        },
        end_point: Point {
            row: 10,
            column: 24,
        },
    },
    Range {
        start_byte: 93,
        end_byte: 134,
        start_point: Point {
            row: 9,
            column: 2,
        },
        end_point: Point {
            row: 11,
            column: 3,
        },
    },
    Range {
        start_byte: 18,
        end_byte: 344,
        start_point: Point {
            row: 2,
            column: 0,
        },
        end_point: Point {
            row: 26,
            column: 1,
        },
    },
    Range {
        start_byte: 0,
        end_byte: 525,
        start_point: Point {
            row: 0,
            column: 0,
        },
        end_point: Point {
            row: 46,
            column: 0,
        },
    },
]
